File: mcp-mahi.py
"""
MAHI Multi-Agent MCP Server for OpenCode
Routes tasks to specialized agents.
"""
import sys
import os
import json

# Fix Windows encoding
if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

# Add MAHI to path
sys.path.insert(0, r"C:\Users\Admin\MAHI")

from router.classifier import classify
from orchestrator.engine import Orchestrator
from agents.base import Task


# Import all agent creators
from agents.code_agent import create_code_agent
from agents.writing_agent import create_writing_agent
from agents.research_agent import create_research_agent
from agents.career_agent import create_career_agent
from agents.teaching_agent import create_teaching_agent
from agents.dss_agent import create_dss_agent
from agents.spiritual_agent import create_spiritual_agent
from agents.quick_agent import create_quick_agent


# Initialize orchestrator
orchestrator = Orchestrator()
for creator in [create_code_agent, create_writing_agent, create_research_agent,
                create_career_agent, create_teaching_agent, create_dss_agent,
                create_spiritual_agent, create_quick_agent]:
    orchestrator.register_agent(creator())


def route_task(user_input: str) -> dict:
    """Classify and route a task."""
    classification = classify(user_input)
    return {
        "category": classification.category,
        "agent": classification.agent_id,
        "model": classification.model,
        "confidence": classification.confidence,
        "urgency": classification.urgency,
    }


def get_agent_status() -> list:
    """Get status of all agents."""
    agents = []
    for agent_id, agent in orchestrator.agents.items():
        agents.append({
            "id": agent.id,
            "name": agent.name,
            "model": agent.config.model_primary,
            "state": agent.state.value,
        })
    return agents


def get_orchestrator_status() -> dict:
    """Get orchestrator status."""
    return {
        "agents": len(orchestrator.agents),
        "queue": len(orchestrator.task_queue),
        "active": len(orchestrator.active_tasks),
        "completed": len(orchestrator.completed_tasks),
    }


# MCP Protocol Handler
def handle_request(request: dict) -> dict:
    """Handle MCP JSON-RPC request."""
    method = request.get("method", "")
    params = request.get("params", {})
    req_id = request.get("id")

    if method == "initialize":
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "result": {
                "protocolVersion": "2024-11-05",
                "capabilities": {"tools": {"listChanged": False}},
                "serverInfo": {"name": "mahi-multi-agent", "version": "1.0.0"},
            },
        }

    elif method == "notifications/initialized":
        return None

    elif method == "tools/list":
        return {
            "jsonrpc": "2.0",
            "id": req_id,
            "result": {
                "tools": [
                    {
                        "name": "mahi_route",
                        "description": "Route a task to the appropriate MAHI agent (code, writing, research, career, teaching, dss, spiritual, quick)",
                        "inputSchema": {
                            "type": "object",
                            "properties": {
                                "input": {"type": "string", "description": "Task description"},
                            },
                            "required": ["input"],
                        },
                    },
                    {
                        "name": "mahi_agents",
                        "description": "List all MAHI agents and their status",
                        "inputSchema": {"type": "object", "properties": {}},
                    },
                    {
                        "name": "mahi_status",
                        "description": "Get MAHI orchestrator status",
                        "inputSchema": {"type": "object", "properties": {}},
                    },
                ]
            },
        }

    elif method == "tools/call":
        tool_name = params.get("name", "")
        arguments = params.get("arguments", {})

        try:
            if tool_name == "mahi_route":
                result = route_task(arguments["input"])
            elif tool_name == "mahi_agents":
                result = get_agent_status()
            elif tool_name == "mahi_status":
                result = get_orchestrator_status()
            else:
                return {"jsonrpc": "2.0", "id": req_id, "error": {"code": -32601, "message": f"Unknown tool: {tool_name}"}}

            return {"jsonrpc": "2.0", "id": req_id, "result": {"content": [{"type": "text", "text": json.dumps(result, indent=2, ensure_ascii=False)}]}}
        except Exception as e:
            return {"jsonrpc": "2.0", "id": req_id, "error": {"code": -32000, "message": str(e)}}

    return {"jsonrpc": "2.0", "id": req_id, "error": {"code": -32601, "message": f"Unknown method: {method}"}}


def main():
    """Run MCP server via stdio."""
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            request = json.loads(line)
            response = handle_request(request)
            if response is not None:
                print(json.dumps(response, ensure_ascii=False))
                sys.stdout.flush()
        except json.JSONDecodeError:
            pass


if __name__ == "__main__":
    main()

File: mcp-mahi.js
#!/usr/bin/env node
/**
 * MAHI Multi-Agent MCP Server for OpenCode
 */
const { execSync } = require("child_process");

function route(input) {
  const text = input.toLowerCase();
  const patterns = [
    [/\b(write|create|build|generate|code|function|class|script|implement)\b.*\b(code|python|javascript|function|class|script)\b/i, "code.generate", "code", 0.85],
    [/\b(review|check|audit|fix|debug|bug|error)\b.*\b(code|script|function)\b/i, "code.review", "code", 0.85],
    [/\b(write|draft|compose|create)\b.*\b(email|message|letter|note)\b/i, "write.email", "write", 0.85],
    [/\b(write|draft|compose|create)\b.*\b(proposal|cover letter|application)\b/i, "write.proposal", "write", 0.85],
    [/\b(search|find|look up|research|investigate)\b/i, "research.find", "research", 0.85],
    [/\b(apply|application|submit|job|position|cv|resume|portfolio|linkedin)\b/i, "career.apply", "career", 0.85],
    [/\b(exercise|lesson|quiz|test|assessment|teach|grammar|vocabulary)\b/i, "teaching.create", "teaching", 0.85],
    [/\b(erp|vba|excel|academix|macro|inventory|stock)\b/i, "dss.work", "dss", 0.85],
    [/\b(astrology|chart|horoscope|natal|transit|nakshatra|quran|surah|spiritual|dhikr)\b/i, "spiritual.read", "spiritual", 0.85],
  ];

  for (const [re, cat, agent, conf] of patterns) {
    if (re.test(text)) {
      const models = { code: "nvidia/nemotron-3-nano-30b-a3b:free", write: "nvidia/nemotron-3-super-120b-a12b:free", research: "nvidia/nemotron-3-super-120b-a12b:free", career: "nvidia/nemotron-3-super-120b-a12b:free", teaching: "nvidia/nemotron-3-nano-30b-a3b:free", dss: "nvidia/nemotron-3-super-120b-a12b:free", spiritual: "google/gemma-4-26b-a4b-it:free" };
      return { category: cat, agent, model: models[agent] || "google/gemma-4-26b-a4b-it:free", confidence: conf, urgency: "normal" };
    }
  }
  return text.split(/\s+/).length < 10
    ? { category: "quick.ask", agent: "quick", model: "google/gemma-4-26b-a4b-it:free", confidence: 0.5, urgency: "instant" }
    : { category: "code.generate", agent: "code", model: "nvidia/nemotron-3-nano-30b-a3b:free", confidence: 0.4, urgency: "normal" };
}

const AGENTS = [
  { id: "code", name: "Code Agent", model: "nvidia/nemotron-3-nano-30b-a3b:free" },
  { id: "write", name: "Writing Agent", model: "nvidia/nemotron-3-super-120b-a12b:free" },
  { id: "research", name: "Research Agent", model: "nvidia/nemotron-3-super-120b-a12b:free" },
  { id: "career", name: "Career Agent", model: "nvidia/nemotron-3-super-120b-a12b:free" },
  { id: "teaching", name: "Teaching Agent", model: "nvidia/nemotron-3-nano-30b-a3b:free" },
  { id: "dss", name: "DSS Agent", model: "nvidia/nemotron-3-super-120b-a12b:free" },
  { id: "spiritual", name: "Spiritual Agent", model: "google/gemma-4-26b-a4b-it:free" },
  { id: "quick", name: "Quick Agent", model: "google/gemma-4-26b-a4b-it:free" },
];

const TOOLS = [
  { name: "mahi_route", description: "Route a task to the appropriate MAHI agent", inputSchema: { type: "object", properties: { input: { type: "string" } }, required: ["input"] } },
  { name: "mahi_agents", description: "List all MAHI agents and their status", inputSchema: { type: "object", properties: {} } },
  { name: "mahi_status", description: "Get MAHI orchestrator status", inputSchema: { type: "object", properties: {} } },
];

function handle(req) {
  const { method, params, id } = req;
  if (method === "initialize") return { jsonrpc: "2.0", id, result: { protocolVersion: "2024-11-05", capabilities: { tools: {} }, serverInfo: { name: "mahi", version: "1.0.0" } } };
  if (method === "notifications/initialized") return null;
  if (method === "tools/list") return { jsonrpc: "2.0", id, result: { tools: TOOLS } };
  if (method === "tools/call") {
    const { name, arguments: args } = params;
    try {
      let result;
      if (name === "mahi_route") result = route(args.input);
      else if (name === "mahi_agents") result = AGENTS;
      else if (name === "mahi_status") result = { agents: AGENTS.length, queue: 0, active: 0, completed: 0 };
      else return { jsonrpc: "2.0", id, error: { code: -32601, message: `Unknown tool: ${name}` } };
      return { jsonrpc: "2.0", id, result: { content: [{ type: "text", text: JSON.stringify(result, null, 2) }] } };
    } catch (e) { return { jsonrpc: "2.0", id, error: { code: -32000, message: e.message } }; }
  }
  return { jsonrpc: "2.0", id, error: { code: -32601, message: `Unknown method: ${method}` } };
}

process.stdin.setEncoding("utf-8");
let buf = "";
process.stdin.on("data", (chunk) => {
  buf += chunk;
  const lines = buf.split("\n");
  buf = lines.pop();
  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    try { const req = JSON.parse(trimmed); const resp = handle(req); if (resp) process.stdout.write(JSON.stringify(resp) + "\n"); } catch {}
  }
});
